### DISCUSSION TITLE
Pimcore 10 installing Azure blob bundle is having issue with routing

### DISCUSSION THREAD

**Discussion by manojthotla25** [pimcore/pimcore#10791] [Q&A / Support]
Hi Team,

I am facing few issues after installing the Azure Bundle for pimcore 10 and i am facing the below  error. However i have installed it in 6.8 version and it is working fine.

I am able see the The Azure Container Configuration button in my settings after installing the bundle but when i click on it i am getting the below error:

![image](https://user-images.githubusercontent.com/93549976/140965685-c03e382f-1639-428e-94fa-1082275d2033.png)

AzureController.php:

<?php

namespace AzurePimcoreBundle\Controller;

use Pimcore\Controller\FrontendController;
use Pimcore\Cache;
use Pimcore\Cache\Core\CoreHandlerInterface;
use Pimcore\Cache\Symfony\CacheClearer;
//use Pimcore\Config;
use Pimcore\Db\Connection;
use Pimcore\Event\SystemEvents;
use Pimcore\File;
use Pimcore\Model;
use Pimcore\Model\Asset;
use Pimcore\Model\Document;
use Pimcore\Model\Element;
use Pimcore\Model\Glossary;
use Pimcore\Model\Metadata;
use Pimcore\Model\Property;
use Pimcore\Model\Staticroute;
use Pimcore\Model\Tool\Tag;
use Pimcore\Model\WebsiteSetting;
use Pimcore\Tool;
use Symfony\Component\EventDispatcher\EventDispatcherInterface;
use Symfony\Component\Filesystem\Filesystem;
use Symfony\Component\HttpFoundation\BinaryFileResponse;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\KernelEvents;
use Symfony\Component\HttpKernel\KernelInterface;
use Symfony\Component\Routing\Annotation\Route;
use AzurePimcoreBundle\Lib\Config;

class AzureSettingController extends FrontendController {

    /**
     * @Route("/admin/settings/get-azure", name="get_azure")
     *
     * @param Request $request
     *
     * @return JsonResponse
     */
    public function getAzureAction(Request $request)
    {
//        $this->checkPermission('azure_settings');

        $values = Config::getAzureConfig();
        $valueArray = $values->toArray();
        $optionsString = [];
        if ($valueArray['azureOptions']) {
            foreach ($valueArray['azureOptions'] as $key => $value) {
                $tmpStr = '--'.$key;
                if ($value !== null && $value !== '') {
                    $tmpStr .= ' '.$value;
                }
                $optionsString[] = $tmpStr;
            }
        }
        $valueArray['azureOptions'] = implode("\n", $optionsString);

        $response = [
            'values' => $valueArray
        ];

        return $this->json($response);
    }
    
    /**
     * @Route("/admin/settings/set-azure", name="set_azure")
     *
     * @param Request $request
     *
     * @return JsonResponse
     */
    public function setAzureAction(Request $request)
    {
//        $this->checkPermission('azure_settings');

        $values = $this->decodeJson($request->get('data'));

        if ($values['azureOptions']) {
            $optionArray = [];
            $lines = explode("\n", $values['azureOptions']);
            foreach ($lines as $line) {
                $parts = explode(' ', substr($line, 2));
                $key = trim($parts[0]);
                if ($key) {
                    $value = trim($parts[1]);
                    $optionArray[$key] = $value;
                }
            }
            $values['azureOptions'] = $optionArray;
        }

        $configFile = \Pimcore\Config::locateConfigFile('azure.php');       
        File::putPhpFile($configFile, to_php_data_file_format($values));

        return $this->json(['success' => true]);
    }
    
    /**
     * Decodes a JSON string into an array/object
     *
     * @param mixed $json       The data to be decoded
     * @param bool $associative Whether to decode into associative array or object
     * @param array $context    Context to pass to serializer when using serializer component
     * @param bool $useAdminSerializer
     *
     * @return array|\stdClass
     */
    protected function decodeJson($json, $associative = true, array $context = [], bool $useAdminSerializer = true)
    {
        /** @var SerializerInterface|DecoderInterface $serializer */
        $serializer = null;

        if ($useAdminSerializer) {
            $serializer = $this->container->get('pimcore_admin.serializer');
        } else {
            $serializer = $this->container->get('serializer');
        }

        if ($associative) {
            $context['json_decode_associative'] = true;
        }

        return $serializer->decode($json, 'json', $context);
    }
    
}
**********************************************************************************************************
Routing.yml:
get_azure:
    path: /admin/settings/get-azure
    defaults: { _controller: AzurePimcoreBundle:AzureSetting:getAzure}
set_azure:
    path: /admin/settings/set-azure
    defaults: { _controller: AzurePimcoreBundle:AzureSetting:setAzure}
*********************************************************************************************************************
azure.js

pimcore.registerNS("pimcore.plugin.azure");

pimcore.plugin.azure = Class.create({
    initialize: function () {
        this.getData();
    },
    getData: function () {
        Ext.Ajax.request({
            url: "/admin/settings/get-azure",
            success: function (response) {

                this.data = Ext.decode(response.responseText);
                this.getTabPanel();

            }.bind(this)
        });
    },
    getValue: function (key, ignoreCheck) {

        var nk = key.split("\.");
        var current = this.data.values;

        for (var i = 0; i < nk.length; i++) {
            if (current[nk[i]]) {
                current = current[nk[i]];
            } else {
                current = null;
                break;
            }
        }

        if (ignoreCheck || (typeof current != "object" && typeof current != "array" && typeof current != "function")) {
            return current;
        }

        return "";
    },
    activate: function () {
        var tabPanel = Ext.getCmp("pimcore_panel_tabs");
        tabPanel.setActiveItem("pimcore_settings_azure");
    },

    save: function () {
        var values = this.layout.getForm().getFieldValues();

        Ext.Ajax.request({
            url: "/admin/settings/set-azure",
            method: "PUT",
            params: {
                data: Ext.encode(values)
            },
            success: function (response) {
                try {
                    var res = Ext.decode(response.responseText);
                    if (res.success) {
                        pimcore.helpers.showNotification(t("success"), t("saved_successfully"), "success");

                        Ext.MessageBox.confirm(t("info"), t("reload_pimcore_changes"), function (buttonValue) {
                            if (buttonValue == "yes") {
                                window.location.reload();
                            }
                        }.bind(this));
                    } else {
                        pimcore.helpers.showNotification(t("error"), t("saving_failed"),
                            "error", t(res.message));
                    }
                } catch (e) {
                    pimcore.helpers.showNotification(t("error"), t("saving_failed"), "error");
                }
            }
        });
    },
    getTabPanel: function () {

        if (!this.panel) {
            this.panel = Ext.create('Ext.panel.Panel', {
                id: "pimcore_settings_azure",
                title: t("Azure Blob Container Settings"),
                iconCls: "pimcore_icon_printpage pimcore_icon_overlay_setting",
                border: false,
                layout: "fit",
                closable: true
            });

            this.panel.on("destroy", function () {
                pimcore.globalmanager.remove("settings_web2print");
            }.bind(this));

            this.pdfReactorSettings = Ext.create("Ext.form.FieldSet", {
                title: t('Azure Blob Container Settings'),
                collapsible: true,
                collapsed: false,
                hidden: this.getValue("generalTool") != 'pdfreactor',
                autoHeight: true,
                defaultType: 'textfield',
                defaults: { width: 450 },
                items: [
                    {
                        fieldLabel: t("web2print_version"),
                        xtype: "combo",
                        width: 600,
                        editable: false,
                        name: "pdfreactorVersion",
                        value: this.getValue("pdfreactorVersion"),
                        store: [
                            ["8.0", "8.0"],
                            ["8.1", "8.1"],
                            ["9.0", "9.0"]
                        ],
                        mode: "local",
                        triggerAction: "all"
                    }, {
                        fieldLabel: t("web2print_protocol"),
                        xtype: "combo",
                        width: 600,
                        editable: false,
                        name: "pdfreactorProtocol",
                        value: this.getValue("pdfreactorProtocol"),
                        store: [
                            ["http", "http"],
                            ["https", "https"]
                        ],
                        mode: "local",
                        triggerAction: "all"
                    }, {
                        xtype: 'textfield',
                        width: 650,
                        fieldLabel: t("web2print_server"),
                        name: 'pdfreactorServer',
                        value: this.getValue("pdfreactorServer")
                    }, {
                        xtype: 'textfield',
                        width: 650,
                        fieldLabel: t("web2print_port"),
                        name: 'pdfreactorServerPort',
                        value: this.getValue("pdfreactorServerPort"),
                        emptyText: "9423"
                    }, {
                        xtype: 'textfield',
                        width: 650,
                        fieldLabel: t("web2print_baseURL"),
                        name: 'pdfreactorBaseUrl',
                        value: this.getValue("pdfreactorBaseUrl")
                    }, {
                        xtype: "displayfield",
                        hideLabel: true,
                        width: 600,
                        value: t('web2print_baseURL_txt'),
                        emptyText: "http://my-domain.org",
                        cls: "pimcore_extra_label_bottom"
                    }, {
                        xtype: 'textfield',
                        width: 650,
                        fieldLabel: t("web2print_apiKey"),
                        name: 'pdfreactorApiKey',
                        value: this.getValue("pdfreactorApiKey")
                    }, {
                        xtype: "displayfield",
                        hideLabel: true,
                        width: 600,
                        value: t('web2print_apiKey_txt'),
                        cls: "pimcore_extra_label_bottom"
                    }, {
                        xtype: 'textarea',
                        width: 650,
                        height: 200,
                        fieldLabel: t("web2print_licence"),
                        name: 'pdfreactorLicence',
                        value: this.getValue("pdfreactorLicence")
                    }
                ]
            });
            this.wkhtmlToPdfSettings = Ext.create("Ext.form.FieldSet", {
                title: t('web2print_wkhtmltopdf_settings'),
                collapsible: true,
                collapsed: false,
                autoHeight: true,
                hidden: this.getValue("generalTool") != 'wkhtmltopdf',
                defaultType: 'textfield',
                defaults: { width: 450 },
                items: [
                    {
                        xtype: 'textfield',
                        width: 650,
                        fieldLabel: t("web2print_wkhtmltopdf_binary"),
                        name: 'wkhtmltopdfBin',
                        value: this.getValue("wkhtmltopdfBin")
                    }, {
                        xtype: 'textarea',
                        width: 650,
                        height: 200,
                        fieldLabel: t("web2print_wkhtmltopdf_options"),
                        name: 'wkhtml2pdfOptions',
                        value: this.getValue("wkhtml2pdfOptions")
                    }, {
                        xtype: "displayfield",
                        hideLabel: true,
                        width: 600,
                        value: t('web2print_wkhtmltopdf_options_txt'),
                        cls: "pimcore_extra_label_bottom"
                    }, {
                        xtype: 'textfield',
                        width: 650,
                        fieldLabel: t("web2print_hostname"),
                        name: 'wkhtml2pdfHostname',
                        value: this.getValue("wkhtml2pdfHostname")
                    }
                ]
            });

            this.layout = Ext.create('Ext.form.Panel', {
                bodyStyle: 'padding:20px 5px 20px 5px;',
                border: false,
                autoScroll: true,
                forceLayout: true,
                defaults: {
                    forceLayout: true
                },
                fieldDefaults: {
                    labelWidth: 150
                },
                buttons: [
                    {
                        text: t("save"),
                        handler: this.save.bind(this),
                        iconCls: "pimcore_icon_apply"
                    }
                ],
                items: [
                    {
                        xtype: 'fieldset',
                        title: t('general'),
                        collapsible: false,
                        autoHeight: true,
                        defaultType: 'textfield',
                        defaults: { width: 650 },
                        items: [
                            {
                                xtype: 'textfield',
                                width: 650,
                                fieldLabel: t("Account URL"),
                                name: 'accountUrl',
                                value: this.getValue("accountUrl")
                            },
                            {
                                xtype: 'textfield',
                                width: 650,
                                fieldLabel: t("Account Name"),
                                name: 'accountName',
                                value: this.getValue("accountName")
                            },
                            {
                                xtype: 'textfield',
                                width: 650,
                                fieldLabel: t("Account Key"),
                                name: 'accountKey',
                                value: this.getValue("accountKey")
                            },
                            {
                                xtype: 'textfield',
                                width: 650,
                                fieldLabel: t("Container"),
                                name: 'container',
                                value: this.getValue("container")
                            },
                            {
                                xtype: 'checkboxfield',
                                name: 'enableAzure',
                                value: this.getValue("enableAzure"),
                                fieldLabel: t("Enable Azure Container?"),
                                width: 650,
                            }

                        ]
                    }

                ]
            });

            this.panel.add(this.layout);

            var tabPanel = Ext.getCmp("pimcore_panel_tabs");
            tabPanel.add(this.panel);
            tabPanel.setActiveItem(this.panel);

            pimcore.layout.refresh();
        }

        return this.panel;
    }
});
************************************************************************************************************
startup.js:

pimcore.registerNS("pimcore.plugin.pimcoreAzureBundle");

pimcore.plugin.pimcoreAzureBundle = Class.create(pimcore.plugin.admin, {
    getClassName: function () {
        return "pimcore.plugin.pimcoreAzureBundle";
    },

    initialize: function () {
        pimcore.plugin.broker.registerPlugin(this);

        Ext.Ajax.timeout = 3600000;
        Ext.Ajax.setTimeout(3600000);
        Ext.override(Ext.data.proxy.Ajax, {timeout: 3600000});        
    },

    pimcoreReady: function (params, broker) {
            var settingsMenu = new Ext.Action({
            text: 'The Azure Container Configurations',
            iconCls: 'pimcore_icon_gridconfig_class_attributes',
            handler: function () {
                var dataExportMainTab = Ext.get("pimcore_settings_azure");
                if (dataExportMainTab) {
                    var tabPanel = Ext.getCmp("pimcore_panel_tabs");
                    tabPanel.setActiveItem("pimcore_settings_azure");
                } else {
                    azuretab = new pimcore.plugin.azure();
                }
            }
        });

        if (layoutToolbar.settingsMenu)
            layoutToolbar.settingsMenu.add(settingsMenu);
    },
});

var AzurePimcoreBundlePlugin = new pimcore.plugin.pimcoreAzureBundle();

Can you please check and let me know if there are anychanges or any other bundle available for pimcore 10 for azure blob:

**Comment by cmann90** [comment]
Hi [manojthotla25](https://github.com/manojthotla25)
the plugin seems to be outdated for the new Pimcore 10 version. The official AWS plugin which the unofficial Azure plugin was based on was replaced by flysystem: https://pimcore.com/docs/pimcore/current/Development_Documentation/Installation_and_Upgrade/System_Setup_and_Hosting/File_Storage_Setup.html 

Offical git repo from **flysystem**: https://github.com/thephpleague/flysystem
Azure Blob support seems to be coming but likely not in Pimcore 10 yet as it was merged 3 days ago (12.02.2022): https://github.com/thephpleague/flysystem/pull/1407

Hope this helps
